### ISSUE TITLE
[3.3.0] Undefined array key 6

### ISSUE THREAD

**Issue by maxstrebel** [spicywebau/craft-neo#639]
#### Description

We are currently testing 3.3.0. If you use "add block above" and add content, you will get an error `Undefined array key 6`

#### Steps to reproduce

1. Create entry
2. Create block
3. Add block above
4. Save

#### Other information
- Neo version: 3.3.0
- Craft version: 4.2.1.1
- (if multi-site) What is the affected Neo field's propagation method:
- Is eager-loading used?

#### Stack trace

```
yii\base\ErrorException: Undefined array key 6 in /var/www/html/vendor/spicyweb/craft-neo/src/validators/FieldValidator.php:158
Stack trace:
#0 /var/www/html/vendor/craftcms/cms/src/web/ErrorHandler.php(77): yii\base\ErrorHandler->handleError(2, 'Undefined array...', '/var/www/html/v...', 158)
#1 /var/www/html/vendor/spicyweb/craft-neo/src/validators/FieldValidator.php(158): craft\web\ErrorHandler->handleError(2, 'Undefined array...', '/var/www/html/v...', 158)
#2 /var/www/html/vendor/spicyweb/craft-neo/src/Field.php(705): benf\neo\validators\FieldValidator->validateAttribute(Object(craft\elements\Entry), 'pageContent')
#3 /var/www/html/vendor/craftcms/cms/src/base/Element.php(2359): benf\neo\Field->validateBlocks(Object(craft\elements\Entry), NULL)
#4 /var/www/html/vendor/yiisoft/yii2/validators/InlineValidator.php(84): craft\base\Element->validateCustomFieldAttribute('field:pageConte...', Array, Object(yii\validators\InlineValidator), Object(benf\neo\elements\db\BlockQuery))
#5 /var/www/html/vendor/yiisoft/yii2/validators/Validator.php(261): yii\validators\InlineValidator->validateAttribute(Object(craft\elements\Entry), 'field:pageConte...')
#6 /var/www/html/vendor/craftcms/cms/src/base/Element.php(2260): yii\validators\Validator->validateAttributes(Object(craft\elements\Entry))
#7 /var/www/html/vendor/yiisoft/yii2/base/Model.php(369): craft\base\Element->afterValidate()
#8 /var/www/html/vendor/craftcms/cms/src/services/Elements.php(2572): yii\base\Model->validate()
#9 /var/www/html/vendor/craftcms/cms/src/services/Elements.php(795): craft\services\Elements->_saveElementInternal(Object(craft\elements\Entry), true, true, NULL)
#10 /var/www/html/vendor/craftcms/cms/src/controllers/ElementsController.php(1262): craft\services\Elements->saveElement(Object(craft\elements\Entry))
#11 [internal function]: craft\controllers\ElementsController->actionApplyDraft()
#12 /var/www/html/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#13 /var/www/html/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#14 /var/www/html/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller->runAction('apply-draft', Array)
#15 /var/www/html/vendor/craftcms/cms/src/web/Application.php(301): yii\base\Module->runAction('elements/apply-...', Array)
#16 /var/www/html/vendor/craftcms/cms/src/web/Application.php(625): craft\web\Application->runAction('elements/apply-...', Array)
#17 /var/www/html/vendor/craftcms/cms/src/web/Application.php(280): craft\web\Application->_processActionRequest(Object(craft\web\Request))
#18 /var/www/html/vendor/yiisoft/yii2/base/Application.php(384): craft\web\Application->handleRequest(Object(craft\web\Request))
#19 /var/www/html/web/index.php(22): yii\base\Application->run()
#20 {main}
```

**Comment by ttempleton** [comment]
I'm unable to reproduce this, and the [relevant code](https://github.com/spicywebau/craft-neo/blob/32b814edb55c33e911bb4af022e7a43d392484cd/src/validators/FieldValidator.php#L143-L158) should be making sure all valid array keys are set.

Can you reproduce this on other Craft installs or on a new install?

**Comment by maxstrebel** [comment]
I will test right away. I also think there's something deeper wrong. When testing in the backend we keep getting the following errors:

- `Invalid Neo block owner ID: 554602`
- `[error] yii\db\Exception: SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction`
- `A provisional draft already exists for element/user` (shows up as error and warning where usually the green checkmark is shown, as also on the logs)

Seemingly at random. I know it is hard to pin down stuff like that. But there is an issue with the drafts. And it is now for weeks breaking my clients backend (for months now).

There's also other tickets pointing in this direction:

- https://github.com/craftcms/cms/issues/9905
- https://github.com/craftcms/cms/issues/9962

Here's a video:

https://user-images.githubusercontent.com/910594/184852520-65d4af01-fb6c-4316-9809-d15d52a7be3d.mov

**Comment by maxstrebel** [comment]
Okay, I cannot reproduce it in a vanilla install. I will try to hunt down what causes is.

**Comment by vebjorn** [comment]
I upgraded to 3.3.0 today as well, and get the same error(s);
When adding a new Neo-block (level 1) and saving, things are OK. If I add A level-1 block, and one or more child-blocks and try to save, I get the Undefined Array-key error (with another number - guessing the number has to do with available child-blocks or something)

**Comment by ttempleton** [comment]
I can force the error to occur, by setting the block type of a block being used at the top level to not be allowed at the top level. Are there any blocks like that in either of your cases @maxstrebel @vebjorn?

**Comment by vebjorn** [comment]
> I can force the error to occur, by setting the block type of a block being used at the top level to not be allowed at the top level. Are there any blocks like that in either of your cases @maxstrebel @vebjorn?

Not in my case, tried a ton of times and combinations by now, completely blank/new entry, fails at the frist neo-block (top level) - also noticed this pops up during autosaving of a draft right after said neo-block is added to the entry.

![Skjermbilde 2022-08-16 kl  12 38 37](https://user-images.githubusercontent.com/4727251/184864003-fe92b805-fdcb-4e99-907a-d40cb3e93549.png)

**Comment by maxstrebel** [comment]
I also cannot confirm. The block from the video (heading) is not allowed on the top level, but also set as a child.

**Comment by vebjorn** [comment]
I can't think of anything else that the fact that the auto-saving of drafts (where I get an error) would be the root-cause here? I have a block named "section" as level-1, and then a block named "column" as level-2.. If I add a "section", and save.. all is well (although I get a auto-save-draft -error), then - if I add a "column" and save again.. all is also well.

If I try to do all this in one operation, stuff goes erronous

**Comment by maxstrebel** [comment]
Would also be my hunch. Looking at the logs, we have a lot of draft related errors.

- `Invalid Neo block owner ID: 554602`
- `A provisional draft already exists for element/user`
- `[error] yii\web\HttpException:400: No draft was identified by the request. at /var/www/html/vendor/craftcms/cms/src/controllers/ElementsController.php:1237`

**Comment by ttempleton** [comment]
Thanks for checking the top level block situation. If possible, could one of you please send your database backup and composer.json/lock files to plugins@spicyweb.com.au and we'll have a look at it.

**Comment by maxstrebel** [comment]
Sent. Did you receive it?

**Comment by ttempleton** [comment]
Thanks @maxstrebel - confirming I've received it.

**Comment by vebjorn** [comment]
I sent database/composer-dumps as well as well @ttempleton

**Comment by ttempleton** [comment]
Confirming I've received that as well, thanks @vebjorn.

**Comment by ttempleton** [comment]
I'm having some issues running either setup, unfortunately, might need to try a different machine... In the meantime, I've just released Neo 3.3.2 which includes a temporary fix for the undefined array key error. It might cause some issues with validation of the new Min Sibling Blocks of This Type block type setting, so it will need to be replaced with a better fix, but hopefully it should be fine for now.

**Comment by maxstrebel** [comment]
Thanks, appreciate your work on this. We did a rollback, as we also experience the other issues (wrong owners, locks, and others – were already present pre 3.3.0).

**Comment by maxstrebel** [comment]
@vebjorn do you happen to use use LJ Dynamic Fields? Or something else that lets you use Twig in fields?

**Comment by maxstrebel** [comment]
I just uninstalled all plugins one by one, and I think LJ Dynamic Fields interferes. I made a vanilla install and after I installed it and built the query from our project, it started to show these errors.

**Comment by vebjorn** [comment]
> I just uninstalled all plugins one by one, and I think LJ Dynamic Fields interferes. I made a vanilla install and after I installed it and built the query from our project, it started to show these errors.

Yeah. I actually do - starting to suspect the same thing

**Comment by maxstrebel** [comment]
Would you share the offending query? In my case it is, when I query a global set by Id.

````
{% set elementSiteId = element.siteId %}
{% set themes = craft.globalSets().handle('theme').siteId(elementSiteId).one().themes ?? [] %}
````

It's crazy. As soon as I add `.siteId(elementSiteId)` it will make NEO behave strangely. The redactor error (https://github.com/spicywebau/craft-neo/issues/642) was also caused by this line. Sorry I did not see this earlier.

**Comment by vebjorn** [comment]
> Would you share the offending query?

Right, well in my case I'm fetching the values of a supertable-field within the same entry as my NEO resides. So I need to get a hold of the "entry id" somehow

`{% set segment = craft.app.request.getSegment(3)|split('-') %}
{% set theEntryId = segment[0] %}
{% for entry in craft.entries.id(theEntryId).all() %}`

Which worked fine in Craft 3, works in all but during draft-save in Craft 4, then things to haywire

**Comment by ttempleton** [comment]
I moved to a different machine and I've been able to add a better fix for this issue that won't affect validation of the Min Sibling Blocks of This Type setting, and that'll be included in the next release.

Regarding any issues with the LJ Dynamic Fields plugin - I'd recommend trying it out with a Matrix field as well and seeing if that causes similar issues. If it doesn't, please feel free to open a new issue for it.

**Comment by ttempleton** [comment]
Just released Neo 3.3.6, with the better fix for this that doesn't affect Min Sibling Blocks of This Type validation.
